aggregator/src/aggregator.rs

```diff
@@ -1883,7 +1919,41 @@ impl VdafOps {
                         tx.put_collection_job(&collection_job),
                         try_join_all(batches.into_iter().map(|(op, batch)| async move {
                             match op {
-                                Operation::Put => tx.put_batch(&batch).await,
+                                Operation::Put => {
+                                    let rslt = tx.put_batch(&batch).await;
+                                    if matches!(
+                                        rslt,
+                                        Err(datastore::Error::MutationTargetAlreadyExists)
+                                    ) {
+                                        // This codepath can be taken due to a quirk of how the
+                                        // Repeatable Read isolation level works. It cannot occur at
+                                        // the Serializable isolation level.
+                                        //
+                                        // For this codepath to be taken, two writers must
+                                        // concurrently choose to write the same batch (by task ID,
+                                        // batch ID, and aggregation parameter), and this batch must
+                                        // not already exist in the datastore.
+                                        //
+                                        // Both writers will receive `None` from the `get_batch`
+                                        // call, and then both will try to `put_batch`. One of the
+                                        // writers will succeed. The other will fail with a unique
+                                        // constraint violation on (task_id, batch_identifier,
+                                        // aggregation_param), since unique constraints are still
+                                        // enforced even in the presence of snapshot isolation. This
+                                        // unique constraint will be translated to a
+                                        // MutationTargetAlreadyExists error.
+                                        //
+                                        // The failing writer, in this case, can't do anything about
+                                        // this problem while in its current transaction: further
+                                        // attempts to read the batch will continue to return `None`
+                                        // (since all reads in the same transaction are from the
+                                        // same snapshot), so it can't update the now-written batch.
+                                        // All it can do is give up on this transaction and try
+                                        // again, by calling `retry` and returning an error.
+                                        tx.retry();
```

Comment:
Naive question: is it not possible to immediately rollback the transaction and return the error?

I otherwise understand how this all comes together.

Reply:
Effectively, that's what will happen: we still return `rslt` from this block, so if we followed this codepath to call `tx.retry()` we will also return an error, which will cause the transaction to be rolled back as quickly as possible (and then retried, since we called `tx.retry()`).

Rollback-and-retry is triggered in two ways:
* Automatically by the datastore transaction if any given datastore operation returns a serialization error (which is how Postgres suggests doing a retry).
* Manually if `tx.retry()` is called. This is necessary because, under Repeatable Read, not every error mode that indicates a serialization failure actually results in a serialization failure. For example, here we would "normally" end up with an error around a conflicting index write; but because we wrote the underlying SQL write query `ON CONFLICT DO NOTHING` and return `MutationTargetAlreadyExists` if the query did not affect any rows, we can get around that -- but we need to retry manually.

(We don't automatically retry in the datastore on seeing the `MutationTargetAlreadyExists` because from the datastore's perspective, it doesn't know if this indicates a concurrent-write conflict or a bug, and if it's the latter automatically retrying might lead to an infinite loop of retries. I'm thinking through if/how we might move things into the datastore, though, as peppering our application code with `retry()` calls does not feel like a fantastic solution.)